datafusion/core/src/physical_plan/offset.rs

```diff
@@ -0,0 +1,316 @@
+        let stream = self.input.execute(0, context)?;
+        Ok(Box::pin(OffsetStream::new(
+            stream,
+            self.offset,
+            baseline_metrics,
+        )))
+    }
+
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default => {
+                write!(f, "OffsetExec: offset={}", self.offset)
+            }
+        }
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn statistics(&self) -> Statistics {
+        let input_stats = self.input.statistics();
+        match input_stats {
+            Statistics {
+                num_rows: Some(nr), ..
+            } => Statistics {
+                num_rows: Some(nr - self.offset),
+                is_exact: input_stats.is_exact,
+                ..Default::default()
+            },
+            _ => Statistics::default(),
+        }
+    }
+}
+
+/// An Offset stream skip the input stream's data up to `offset` row.
+pub struct OffsetStream {
+    /// Number of rows to skip, starts with 1.
+    offset: usize,
+    input: SendableRecordBatchStream,
+    schema: SchemaRef,
+    /// Number of rows have already skipped.
+    skipped: usize,
+    /// Execution time metrics
+    baseline_metrics: BaselineMetrics,
+}
+
+impl OffsetStream {
+    fn new(
+        input: SendableRecordBatchStream,
+        offset: usize,
+        baseline_metrics: BaselineMetrics,
+    ) -> Self {
+        let schema = input.schema();
+        Self {
+            offset,
+            input,
+            schema,
+            skipped: 0,
+            baseline_metrics,
+        }
+    }
+
+    fn poll_and_skip(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<ArrowResult<RecordBatch>>> {
+        loop {
+            let poll = self.input.poll_next_unpin(cx);
+            let poll = poll.map_ok(|batch| {
+                if batch.num_rows() + self.skipped <= self.offset {
+                    self.skipped += batch.num_rows();
+                    RecordBatch::new_empty(self.input.schema())
+                } else {
+                    let new_batch = cut_batch(&batch, self.offset - self.skipped);
+                    self.skipped = self.offset;
+                    new_batch
+                }
+            });
+
+            match &poll {
+                Poll::Ready(Some(Ok(batch)))
+                    if batch.num_rows() > 0 && self.skipped == self.offset =>
+                {
+                    break poll
+                }
+                Poll::Ready(Some(Err(_e))) => break poll,
+                Poll::Ready(None) => break poll,
+                Poll::Pending => break poll,
+                _ => {
+                    // continue to poll input stream
+                }
+            }
+        }
+    }
+}
+
+/// Remove a RecordBatch's first n rows
+pub fn cut_batch(batch: &RecordBatch, n: usize) -> RecordBatch {
```

Comment:
I think you might be able to use `RecordBatch::slice()` here

https://docs.rs/arrow/14.0.0/arrow/record_batch/struct.RecordBatch.html#method.slice